// @react-native-community/cli-tools/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  logger: true,
  isPackagerRunning: true,
  getDefaultUserTerminal: true,
  fetch: true,
  fetchToTemp: true,
  launchDefaultBrowser: true,
  launchDebugger: true,
  launchEditor: true,
  releaseChecker: true,
  resolveNodeModuleDir: true,
  hookStdout: true,
  getLoader: true,
  NoopLoader: true,
  Loader: true,
  findProjectRoot: true
};
Object.defineProperty(exports, "Loader", {
  enumerable: true,
  get: function () {
    return _loader.Loader;
  }
});
Object.defineProperty(exports, "NoopLoader", {
  enumerable: true,
  get: function () {
    return _loader.NoopLoader;
  }
});
Object.defineProperty(exports, "fetch", {
  enumerable: true,
  get: function () {
    return _fetch.fetch;
  }
});
Object.defineProperty(exports, "fetchToTemp", {
  enumerable: true,
  get: function () {
    return _fetch.fetchToTemp;
  }
});
Object.defineProperty(exports, "findProjectRoot", {
  enumerable: true,
  get: function () {
    return _findProjectRoot.default;
  }
});
Object.defineProperty(exports, "getDefaultUserTerminal", {
  enumerable: true,
  get: function () {
    return _getDefaultUserTerminal.default;
  }
});
Object.defineProperty(exports, "getLoader", {
  enumerable: true,
  get: function () {
    return _loader.getLoader;
  }
});
Object.defineProperty(exports, "hookStdout", {
  enumerable: true,
  get: function () {
    return _hookStdout.default;
  }
});
Object.defineProperty(exports, "isPackagerRunning", {
  enumerable: true,
  get: function () {
    return _isPackagerRunning.default;
  }
});
Object.defineProperty(exports, "launchDebugger", {
  enumerable: true,
  get: function () {
    return _launchDebugger.default;
  }
});
Object.defineProperty(exports, "launchDefaultBrowser", {
  enumerable: true,
  get: function () {
    return _launchDefaultBrowser.default;
  }
});
Object.defineProperty(exports, "launchEditor", {
  enumerable: true,
  get: function () {
    return _launchEditor.default;
  }
});
Object.defineProperty(exports, "logger", {
  enumerable: true,
  get: function () {
    return _logger.default;
  }
});
Object.defineProperty(exports, "releaseChecker", {
  enumerable: true,
  get: function () {
    return _releaseChecker.default;
  }
});
Object.defineProperty(exports, "resolveNodeModuleDir", {
  enumerable: true,
  get: function () {
    return _resolveNodeModuleDir.default;
  }
});
var _logger = _interopRequireDefault(require("./logger"));
var _isPackagerRunning = _interopRequireDefault(require("./isPackagerRunning"));
var _getDefaultUserTerminal = _interopRequireDefault(require("./getDefaultUserTerminal"));
var _fetch = require("./fetch");
var _launchDefaultBrowser = _interopRequireDefault(require("./launchDefaultBrowser"));
var _launchDebugger = _interopRequireDefault(require("./launchDebugger"));
var _launchEditor = _interopRequireDefault(require("./launchEditor"));
var _releaseChecker = _interopRequireDefault(require("./releaseChecker"));
var _resolveNodeModuleDir = _interopRequireDefault(require("./resolveNodeModuleDir"));
var _hookStdout = _interopRequireDefault(require("./hookStdout"));
var _loader = require("./loader");
var _findProjectRoot = _interopRequireDefault(require("./findProjectRoot"));
var _errors = require("./errors");
Object.keys(_errors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _errors[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _errors[key];
    }
  });
});
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
