// @shopify/react-native-skia/lib/commonjs/renderer/Reconciler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SkiaRoot = void 0;

var _reactReconciler = _interopRequireDefault(require("react-reconciler"));

var _DependencyManager = require("./DependencyManager");

var _HostConfig = require("./HostConfig");

var _Container = require("./Container");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const skiaReconciler = (0, _reactReconciler.default)(_HostConfig.skHostConfig);

const createDependencyManager = registerValues => global.SkiaDomApi && global.SkiaDomApi.DependencyManager ? global.SkiaDomApi.DependencyManager(registerValues) : new _DependencyManager.DependencyManager(registerValues);

skiaReconciler.injectIntoDevTools({
  bundleType: 1,
  version: "0.0.1",
  rendererPackageName: "react-native-skia"
});

class SkiaRoot {
  constructor(Skia) {
    let registerValues = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : () => () => {};
    let redraw = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : () => {};

    _defineProperty(this, "root", void 0);

    _defineProperty(this, "container", void 0);

    const depMgr = createDependencyManager(registerValues);
    this.container = new _Container.Container(Skia, depMgr, redraw);
    this.root = skiaReconciler.createContainer(this.container, 0, null, true, null, "", console.error, null);
  }

  render(element) {
    skiaReconciler.updateContainer(element, this.root, null, () => {
      (0, _HostConfig.debug)("updateContainer");
      this.container.depMgr.update();
    });
  }

  unmount() {
    skiaReconciler.updateContainer(null, this.root, null, () => {
      this.container.depMgr.remove();
    });
  }

  get dom() {
    return this.container.root;
  }

}

exports.SkiaRoot = SkiaRoot;
